regrippy/plugins/localgroups.py

```diff
@@ -0,0 +1,144 @@
+import struct
+
+from regrippy import BasePlugin, PluginResult, mactime
+
+
+class Plugin(BasePlugin):
+    """Extract local group information from the SAM & SOFTWARE databases"""
+
+    __REGHIVE__ = ["SOFTWARE", "SAM"]
+
+    __user_profile_list = []
+
+    def sid2asc(self, bin_sid):
+        """Turns a binary SID into its ASCII representation"""
+        # SID structure:
+        #   + 1 byte (revision, little endian)
+        #   + 1 byte (count of sub auths, little endian)
+        #   + 6 bytes (authority value, big endian)
+        #   + n * 4 bytes where n is the count of sub auths
+        len_sid = len(bin_sid)
+        if len_sid == 12:
+            rev, _, bytes_idauth, sub = struct.unpack("<BB6sL", bin_sid)
+            int_idauth = int.from_bytes(bytes_idauth, byteorder="big", signed=False)
+            return "S-{0}-{1}-{2}".format(rev, int_idauth, sub)
+        elif len_sid == 24:
+            rev, _, bytes_idauth, sub1, sub2, sub3, sub4 = struct.unpack("<BB6s4L", bin_sid)
+            int_idauth = int.from_bytes(bytes_idauth, byteorder="big", signed=False)
+            return "S-{0}-{1}-{2}-{3}-{4}-{5}".format(rev, int_idauth, sub1, sub2, sub3, sub4)
+        elif len_sid == 28:
+            rev, _, bytes_idauth, sub1, sub2, sub3, sub4, rid = struct.unpack("<BB6s4LL", bin_sid)
+            int_idauth = int.from_bytes(bytes_idauth, byteorder="big", signed=False)
+            return "S-{0}-{1}-{2}-{3}-{4}-{5}-{6}".format(rev, int_idauth, sub1, sub2, sub3, sub4, rid)
+        else:
+            raise ValueError("Unsupported SID length (expected  12, 24, or 28): %i" % len_sid)
+
+    def machine_sid(self):
+        """Gets the machine SID from the SAM hive"""
+        # The machine SID on which all local accounts are based sits at the end of the Account key's V data
+        key_machine = self.open_key("SAM\\Domains\\Account")
+        v_raw = key_machine.value("V").value()
+        v_data = v_raw[len(v_raw) - 24 :]
+        return self.sid2asc(v_data)
+
+    def user_sids_sam(self):
+        """Gets user SIDs from the SAM hive"""
+        machine_sid = self.machine_sid()
+        key_accounts = self.open_key("SAM\\Domains\\Account\\Users\\Names")
+        if not key_accounts:
+            return
+        for subkey in key_accounts.subkeys():
+            # User SIDs are built by appending their Relative ID to the machine SID.
+            # Relative IDs are stored inside Names' subkeys; each has a default value where the value *type* holds the
+            # integer value of the RID.
+            sid = "{0}-{1}".format(machine_sid, subkey.value("").value_type())
```

Comment:
Small nitpick: could you use `.value("(default)")` when getting the default value of a key? I find it more explicit.

Reply:
Fixed in 641000a